### PR TITLE
Adjust JWT and SAML fetch-and-update user to save new attributes

### LINKED ISSUE #15484: SSO logins doesn't update user name in Metabase
**Describe the bug**
When users login for the first time via SAML or JWT (looks like it's the same for LDAP and Google), then a Metabase user is created with the details (name, etc) and attributes (groups, etc) from the SSO.
But on subsequent logins, only the attributes are updated - not the name, which means it manually needs to be edited in Metabase to keep it up-to-date.

**To Reproduce**
1. Setup Metabase with SSO
2. Login with a user via SSO, now a user in Metabase is created
3. Change the name of the user in the IdP
4. Login with the same user as step 2, the user's name is not updated

**Expected behavior**
User information received from SSO during creation should also be kept up-to-date from the SSO.
**Note**: When the name changes, it should update the Personal Collection name too - see PR #11379

**Information about your Metabase Installation:**
Metabase 1.35.4.1 thru 1.38.3

**Additional context**
Originally EE477
If wrong first name and last name is configured in the JWT settings, then the name is `Unknown Unknown`, and would need to be manually corrected in Metabase.
Related to #15483

### PR DESCRIPTION
**Pull request by adam-james-v**
Related: #22754
Fixes: #15484

Before this change, JWT/SAML logins would attempt to update attributes, but never considered the first-name or last-name attributes, so they would never be updated, even if the identity provider sends the correct keys.

Now, `:first_name` and `:last_name` attributes are properly considered when calling `sso-utils/fetch-and-update-login-attributes!` which will:

- return the User unchanged if the IdP keys/vals are equal to those in the app-db
- return the User unchanged if the IdP is missing first/last name attributes (first/last name are required by the app-db, so will already exist either as "Unknown" or some previously established value)
- update the User attributes (including `:first_name` and `:last_name`) in the app-db if the IdP attributes do NOT match.

In other words, the IdP values will take precedence over Metabase's app-db values, which was already true before this PR except, of course, for first/last name attributes.

### CODE REVIEW

File: test_resources/saml-test-response-new-user-no-names.xml [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      </saml:Attribute>
+      <saml:Attribute Name="http://schemas.auth0.com/updated_at" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
+        <saml:AttributeValue xsi:type="xs:anyType">Thu Jul 05 2018 18:02:53 GMT+0000 (UTC)</saml:AttributeValue>
+      </saml:Attribute>
+      <saml:Attribute Name="http://schemas.auth0.com/created_at" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
+        <saml:AttributeValue xsi:type="xs:anyType">Tue May 29 2018 22:11:36 GMT+0000 (UTC)</saml:AttributeValue>
+      </saml:Attribute>
+    </saml:AttributeStatement>
+  </saml:Assertion>
+</samlp:Response>
```
Comment (adam-james-v) [inline test_resources/saml-test-response-new-user-no-names.xml:87]: This is equivalent to `saml-test-response-new-user.xml` except given name, surname, and nickname attributes are removed.

This is added for testing the username updating successfully for SAML.

File: enterprise/backend/test/metabase_enterprise/sso/integrations/saml_test.clj [hunk 2]
```diff
@@ -408,6 +413,44 @@
...
+                       :is_superuser false
+                       :id           true
+                       :last_name    "User"
+                       :date_joined  true
+                       :common_name  "New User"}]
+                     (->> (mt/boolean-ids-and-timestamps (db/select User :email "newuser@metabase.com"))
+                          (map #(dissoc % :last_login))))))
+            (finally
+              (db/delete! User :%lower.email "newuser@metabase.com"))))))))
+
```
Comment (adam-james-v) [inline enterprise/backend/test/metabase_enterprise/sso/integrations/saml_test.clj:453]: This test:

1. Makes a new user (which does NOT have `:first_name` `:last_name`)
2. Checks that this newly created user has "Unknown" first/last name
3. Updates the user via an SSO login (a different SAML .xml file is loaded, which is the same user, but has first/last name attributes this time).
4. Check that the first/last name have been changed to match those provided by the identity provider.

File: enterprise/backend/src/metabase_enterprise/sso/integrations/sso_utils.clj [hunk 3]
```diff
@@ -36,13 +36,15 @@
...
-  (when-let [{:keys [id login_attributes] :as user} (db/select-one User :%lower.email (u/lower-case-en email))]
-    (if (= login_attributes new-user-attributes)
-      user
-      (do
-        (db/update! User id :login_attributes new-user-attributes)
-        (User id)))))
+  [{:keys [email] :as user-attributes}]
+  (when-let [{:keys [id] :as user} (db/select-one User :%lower.email (u/lower-case-en email))]
+    (let [attribute-keys (keys user-attributes)
+          user-attributes (into {} (filter second user-attributes))] ;; remove attributes with `nil` values
```
[the remark this answers is not in the document]
Comment (noahmoss) [reply enterprise/backend/src/metabase_enterprise/sso/integrations/sso_utils.clj:42]: I think essentially the same logic is already implemented for LDAP in `metabase-enterprise.enhancements.integrations.ldap/attribute-synced-user`, though the code looks very different. Any chance we could have a single implementation for this, shared by LDAP, JWT and SAML? (I don't really care which implementation you choose, but it would be nice to not have the duplication if possible. And if you can verify that the behavior for LDAP doesn't change.)

File: enterprise/backend/test/metabase_enterprise/sso/integrations/jwt_test.clj [hunk 4]
```diff
@@ -30,7 +30,7 @@
 (def ^:private default-jwt-secret   (crypto-random/hex 32))
 
 (deftest sso-prereqs-test
-  (testing "SSO requests fail if SAML hasn't been enabled"
+  (testing "SSO requests fail if JWT hasn't been enabled"
```
Comment (noahmoss) [inline enterprise/backend/test/metabase_enterprise/sso/integrations/jwt_test.clj:33]: lol 👍

**Review by noahmoss — APPROVED** [review]
Nice, looks correct as far as I can tell. It'd be really cool if we could consolidate the LDAP implementation into this too but it's not strictly necessary, so stamping to unblock...if you end up changing a bunch of LDAP code let me know and I can take another look.

File: enterprise/backend/src/metabase_enterprise/sso/integrations/sso_utils.clj [hunk 5]
```diff
@@ -36,13 +36,15 @@
...
+      (if (= (select-keys user attribute-keys) user-attributes)
+        user
+        (do
+          (apply db/update! (concat [User id] (apply concat user-attributes)))
```
[the remark this answers is not in the document]
Comment (adam-james-v) [reply enterprise/backend/src/metabase_enterprise/sso/integrations/sso_utils.clj]: Haha, yep. Bit of a brain fart there. Thanks :)

File: enterprise/backend/src/metabase_enterprise/sso/integrations/sso_utils.clj [hunk 6]
```diff
@@ -36,13 +36,15 @@
...
 (defn fetch-and-update-login-attributes!
   "Update the login attributes for the user at `email`. This call is a no-op if the login attributes are the same"
-  [email new-user-attributes]
-  (when-let [{:keys [id login_attributes] :as user} (db/select-one User :%lower.email (u/lower-case-en email))]
-    (if (= login_attributes new-user-attributes)
-      user
-      (do
-        (db/update! User id :login_attributes new-user-attributes)
-        (User id)))))
+  [{:keys [email] :as user-attributes}]
```
[the remark this answers is not in the document]
Comment (adam-james-v) [reply enterprise/backend/src/metabase_enterprise/sso/integrations/sso_utils.clj]: This was a really good point, thanks. I went with `user-from-sso` and `user-data`

File: enterprise/backend/src/metabase_enterprise/sso/integrations/sso_utils.clj (hunk 3 above)
[the remark this answers is not in the document]
Comment (adam-james-v) [reply enterprise/backend/src/metabase_enterprise/sso/integrations/sso_utils.clj]: I think LDAP may benefit from a more thorough cleanup to be honest, but I've now altered things a bit to use the `sso-utils/fetch-and-update-user!` function, so we're getting closer to this unification :)

[the remark this answers is not in the document]
Comment (adam-james-v) [reply enterprise/backend/src/metabase_enterprise/sso/integrations/sso_utils.clj]: I've actually reverted the LDAP change. I got stuck and want to move this PR forward, so I'll address LDAP in a follow on PR.
